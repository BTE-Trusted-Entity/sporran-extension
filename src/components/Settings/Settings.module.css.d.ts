declare namespace SettingsModuleCssNamespace {
  export interface ISettingsModuleCss {
    toggle: string;
  }
}

declare const SettingsModuleCssModule: SettingsModuleCssNamespace.ISettingsModuleCss & {
  /** WARNING: Only available when `css-loader` is used without `style-loader` or `mini-css-extract-plugin` */
  locals: SettingsModuleCssNamespace.ISettingsModuleCss;
};

export = SettingsModuleCssModule;
