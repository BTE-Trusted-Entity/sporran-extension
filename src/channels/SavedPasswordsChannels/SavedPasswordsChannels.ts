import { BrowserChannel } from '../base/BrowserChannel/BrowserChannel';
import { channelsEnum } from '../base/channelsEnum';

interface SavedPassword {
  password: string;
  timestamp: number;
}

const savedPasswords: Record<string, SavedPassword> = {};

interface SavePasswordInput {
  password: string;
  address: string;
}

export const savePasswordChannel = new BrowserChannel<SavePasswordInput>(
  channelsEnum.savePassword,
);

export async function savePassword({
  address,
  password,
}: SavePasswordInput): Promise<void> {
  savedPasswords[address] = { password, timestamp: Date.now() };
}

export const getPasswordChannel = new BrowserChannel<
  string,
  string | undefined
>(channelsEnum.getPassword);

export async function getPassword(
  address: string,
): Promise<string | undefined> {
  return savedPasswords[address]?.password;
}

export const forgetPasswordChannel = new BrowserChannel<string>(
  channelsEnum.forgetPassword,
);

export async function forgetPassword(address: string): Promise<void> {
  delete savedPasswords[address];
}

export const hasSavedPasswordsChannel = new BrowserChannel<void, boolean>(
  channelsEnum.hasSavedPasswords,
);

export async function hasSavedPasswords(): Promise<boolean> {
  return Object.values(savedPasswords).length > 0;
}

export const forgetAllPasswordsChannel = new BrowserChannel(
  channelsEnum.forgetAllPasswords,
);

export async function forgetAllPasswords(): Promise<void> {
  for (const password in savedPasswords) {
    delete savedPasswords[password];
  }
}

const saveDuration = 15 * 60 * 1000;
const intervalDuration = 1 * 60 * 1000;

function checkExpiredPasswords(): void {
  const oldestPossible = Date.now() - saveDuration;
  for (const password in savedPasswords) {
    const { timestamp } = savedPasswords[password];
    if (timestamp < oldestPossible) {
      delete savedPasswords[password];
    }
  }
}

export function schedulePasswordsCheck(): void {
  setInterval(checkExpiredPasswords, intervalDuration);
}
