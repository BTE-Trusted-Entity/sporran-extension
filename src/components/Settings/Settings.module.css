.toggle {
  composes: toggle from '../../components/Menu/Menu.module.css';

  display: inline-flex;
  width: 1.625rem;
  height: 1.625rem;
  border: 0;
  padding: 0;
  background: url('../../components/images/settings.svg') no-repeat center;
}
