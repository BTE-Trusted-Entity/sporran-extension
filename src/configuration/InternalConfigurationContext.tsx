import { ConfigurationContext } from './ConfigurationContext';
import { configuration, internalFeatures } from './configuration';

const internalConfiguration = {
  ...configuration,
  features: internalFeatures,
};

export function InternalConfigurationContext({
  children,
}: {
  children: JSX.Element;
}): JSX.Element {
  return (
    <ConfigurationContext.Provider value={internalConfiguration}>
      {children}
    </ConfigurationContext.Provider>
  );
}
