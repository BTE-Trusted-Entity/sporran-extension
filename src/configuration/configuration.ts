import { isInternal } from './variant';

interface Features {
  endpoint: boolean;
  recipientsList: boolean;
  finalized: boolean;
}

export interface ConfigurationType {
  version: string;
  features: Features;
}

const publicFeatures: Features = {
  endpoint: false,
  recipientsList: false,
  finalized: true,
};

export const internalFeatures: Features = {
  endpoint: true,
  recipientsList: true,
  finalized: false,
};

// Duplicates the value in src/static/manifest.json
// We can’t use browser.runtime.getManifest().version, as it’s unavailable in injected scripts
const version = '2023.4.12';

export const configuration: ConfigurationType = {
  version,
  features: isInternal ? internalFeatures : publicFeatures,
};
