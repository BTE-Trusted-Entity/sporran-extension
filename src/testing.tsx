export * from '@testing-library/react';
export { waitForElementToBeRemoved } from '@testing-library/dom';
import { act, render as externalRender, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import dialogPolyfill from 'dialog-polyfill';

import { AccountsProviderMock } from './testing/AccountsProviderMock';

export {
  AccountsProviderMock,
  accountsMock,
} from './testing/AccountsProviderMock';

export { mockBackgroundScript } from './testing/mockBackgroundScript';

export function render(
  ui: Parameters<typeof externalRender>[0],
  options?: Parameters<typeof externalRender>[1],
): ReturnType<typeof externalRender> {
  return externalRender(
    <AccountsProviderMock>
      <MemoryRouter>{ui}</MemoryRouter>
    </AccountsProviderMock>,
    options,
  );
}

export async function waitForTooltipUpdate(): Promise<void> {
  expect(await screen.findByRole('status')).toHaveAttribute(
    'data-popper-escaped',
  );
}

const dialogPromise = Promise.resolve();
(dialogPolyfill.registerDialog as jest.Mock).mockReturnValue(dialogPromise);

export async function waitForDialogUpdate(): Promise<void> {
  await act(() => dialogPromise);
}
