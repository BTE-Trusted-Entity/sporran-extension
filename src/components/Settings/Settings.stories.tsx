import { Meta } from '@storybook/react';
import { MemoryRouter } from 'react-router-dom';

import * as menuStyles from '../Menu/Menu.module.css';

import { IdentitiesProviderMock } from '../../utilities/identities/IdentitiesProvider.mock';

import { Settings } from './Settings';

export default {
  title: 'Components/Settings',
  component: Settings,
  decorators: [
    (Story) => (
      <div className={menuStyles.wrapper} style={{ float: 'right' }}>
        <Story />
      </div>
    ),
  ],
} as Meta;

export function NoIdentities(): JSX.Element {
  return (
    <IdentitiesProviderMock identities={{}}>
      <Settings />
    </IdentitiesProviderMock>
  );
}

export function WithIdentities(): JSX.Element {
  return (
    <MemoryRouter
      initialEntries={[
        '/identity/4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1',
      ]}
    >
      <Settings />
    </MemoryRouter>
  );
}
