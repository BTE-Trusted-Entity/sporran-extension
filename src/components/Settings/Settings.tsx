import { useCallback, useEffect, useState } from 'react';
import { browser } from 'webextension-polyfill-ts';
import useDropdownMenu from 'react-accessible-dropdown-menu-hook';
import { Link, useRouteMatch } from 'react-router-dom';

import * as menuStyles from '../Menu/Menu.module.css';
import * as styles from './Settings.module.css';

import { useIdentities } from '../../utilities/identities/identities';
import {
  forgetAllPasswordsChannel,
  hasSavedPasswordsChannel,
} from '../../channels/SavedPasswordsChannels/SavedPasswordsChannels';
import { useConfiguration } from '../../configuration/useConfiguration';
import { generatePath, paths } from '../../views/paths';

function useItemsCount(onExistingIdentity: boolean, hasPasswords: boolean) {
  const itemCounts = {
    generic: 5,
    identity: onExistingIdentity ? 2 : 0,
    forgetPasswords: onExistingIdentity && hasPasswords ? 1 : 0,
  };

  return itemCounts.generic + itemCounts.identity + itemCounts.forgetPasswords;
}

export function Settings(): JSX.Element {
  const t = browser.i18n.getMessage;

  const [hasPasswords, setHasPasswords] = useState(false);

  const match = useRouteMatch(paths.identity.overview);
  const address = (match?.params as { address: string })?.address;
  const identities = useIdentities().data;
  const onExistingIdentity = Boolean(identities?.[address]);

  const { version } = useConfiguration();

  const count = useItemsCount(onExistingIdentity, hasPasswords);
  const { buttonProps, itemProps, isOpen, setIsOpen } = useDropdownMenu(count);

  const handleClick = useCallback(() => {
    setIsOpen(false);
  }, [setIsOpen]);

  useEffect(() => {
    (async () => {
      setHasPasswords(await hasSavedPasswordsChannel.get());
    })();
  }, [isOpen]);

  const handleForgetAllClick = useCallback(async () => {
    await forgetAllPasswordsChannel.get();
  }, []);

  return (
    <div className={menuStyles.wrapper}>
      <button
        {...buttonProps}
        type="button"
        className={styles.toggle}
        title={t('component_Settings_label')}
        aria-label={t('component_Settings_label')}
      />

      {isOpen && (
        <div className={menuStyles.dropdown} role="menu" onClick={handleClick}>
          <h4 className={menuStyles.heading}>
            {t('component_Settings_label')}
          </h4>

          <ul className={menuStyles.list}>
            {onExistingIdentity && (
              <>
                <li className={menuStyles.listItem}>
                  <Link
                    to={generatePath(paths.identity.remove, { address })}
                    {...itemProps.shift()}
                  >
                    {t('component_Settings_forget')}
                  </Link>
                </li>

                <li className={menuStyles.listItem}>
                  <Link
                    to={generatePath(paths.identity.reset.start, {
                      address,
                    })}
                    {...itemProps.shift()}
                  >
                    {t('component_Settings_reset_password')}
                  </Link>
                </li>
                <li
                  className={
                    hasPasswords ? menuStyles.listItem : menuStyles.disabled
                  }
                >
                  <button
                    type="button"
                    className={menuStyles.listButton}
                    {...(hasPasswords &&
                      (itemProps.shift() as Record<string, unknown>))}
                    onClick={handleForgetAllClick}
                    disabled={!hasPasswords}
                  >
                    {t('component_Settings_forget_saved_passwords')}
                  </button>
                </li>
              </>
            )}

            <li className={menuStyles.listItem}>
              <a
                href="https://www.sporran.org/terms"
                target="_blank"
                rel="noreferrer"
                {...itemProps.shift()}
              >
                {t('component_Settings_terms_and_conditions')}
              </a>
            </li>

            <li className={menuStyles.listItem}>
              <a
                href="https://support.kilt.io/"
                target="_blank"
                rel="noreferrer"
                {...itemProps.shift()}
              >
                {t('component_Settings_tech_support')}
              </a>
            </li>

            <li className={menuStyles.listItem}>
              <Link to={paths.access} {...itemProps.shift()}>
                {t('component_Settings_access')}
              </Link>
            </li>

            <li className={menuStyles.listItem}>
              <Link to={paths.settings} {...itemProps.shift()}>
                {t('component_Settings_endpoint')}
              </Link>
            </li>

            <li className={menuStyles.listItem}>
              <a {...itemProps.shift()}>
                {t('component_Settings_version', [version])}
              </a>
            </li>
          </ul>
        </div>
      )}
    </div>
  );
}
