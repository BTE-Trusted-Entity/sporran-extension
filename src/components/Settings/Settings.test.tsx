import userEvent from '@testing-library/user-event';
import { MemoryRouter } from 'react-router-dom';

import { render, screen } from '../../testing/testing';
import { IdentitiesProviderMock } from '../../utilities/identities/IdentitiesProvider.mock';
import { waitForHasSavedPasswords } from '../../channels/SavedPasswordsChannels/SavedPasswordsChannels.mock';
import { InternalConfigurationContext } from '../../configuration/InternalConfigurationContext';

import { Settings } from './Settings';

describe('Settings', () => {
  it('should render', async () => {
    const { container } = render(<Settings />);
    await waitForHasSavedPasswords();
    expect(container).toMatchSnapshot();
  });

  it('menu should be visible when menu button clicked', async () => {
    const { container } = render(
      <MemoryRouter
        initialEntries={[
          '/identity/4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1',
        ]}
      >
        <Settings />
      </MemoryRouter>,
    );
    const openMenuButton = await screen.findByLabelText('Settings');
    expect(openMenuButton).toHaveAttribute('aria-expanded', 'false');
    expect(screen.queryByRole('menu')).not.toBeInTheDocument();

    await userEvent.click(openMenuButton);

    expect(await screen.findByRole('menu')).toBeInTheDocument();
    expect(
      await screen.findByRole('menuitem', { name: 'Remove Current Identity' }),
    ).toBeInTheDocument();
    expect(openMenuButton).toHaveAttribute('aria-expanded', 'true');

    await waitForHasSavedPasswords();

    expect(container).toMatchSnapshot();
  });

  it('should not render identity options when creating identity', async () => {
    render(
      <MemoryRouter initialEntries={['/identity/NEW']}>
        <Settings />
      </MemoryRouter>,
    );
    await userEvent.click(await screen.findByLabelText('Settings'));

    expect(
      screen.queryByRole('menuitem', { name: 'Forget current identity' }),
    ).not.toBeInTheDocument();
    expect(
      screen.queryByRole('menuitem', {
        name: 'Reset password for current identity',
      }),
    ).not.toBeInTheDocument();

    await waitForHasSavedPasswords();
  });

  it('should not render identity options if there are no identities', async () => {
    render(
      <IdentitiesProviderMock identities={{}}>
        <Settings />
      </IdentitiesProviderMock>,
    );
    await userEvent.click(await screen.findByLabelText('Settings'));

    expect(
      screen.queryByRole('menuitem', { name: 'Forget current identity' }),
    ).not.toBeInTheDocument();
    expect(
      screen.queryByRole('menuitem', {
        name: 'Reset password for current identity',
      }),
    ).not.toBeInTheDocument();

    await waitForHasSavedPasswords();
  });

  it('should render the endpoint item in the internal build', async () => {
    const { container } = render(
      <InternalConfigurationContext>
        <Settings />
      </InternalConfigurationContext>,
    );
    await userEvent.click(await screen.findByLabelText('Settings'));

    expect(
      await screen.findByRole('menuitem', { name: 'Custom Endpoint' }),
    ).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });
});
