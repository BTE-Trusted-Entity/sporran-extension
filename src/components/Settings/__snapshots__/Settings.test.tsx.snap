// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings menu should be visible when menu button clicked 1`] = `
<div>
  <div
    class="wrapper"
  >
    <button
      aria-expanded="true"
      aria-haspopup="true"
      aria-label="Settings"
      class="toggle"
      role="button"
      tabindex="0"
      title="Settings"
      type="button"
    />
    <div
      class="dropdown"
      role="menu"
    >
      <h4
        class="heading"
      >
        Settings
      </h4>
      <ul
        class="list"
      >
        <li
          class="listItem"
        >
          <a
            href="/identity/4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1/remove"
            role="menuitem"
            tabindex="-1"
          >
            Remove Current Identity
          </a>
        </li>
        <li
          class="listItem"
        >
          <a
            href="/identity/4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1/reset"
            role="menuitem"
            tabindex="-1"
          >
            Reset Password for Current Identity
          </a>
        </li>
        <li
          class="disabled"
        >
          <button
            class="listButton"
            disabled=""
            type="button"
          >
            Require Password for Future Transactions
          </button>
        </li>
        <li
          class="listItem"
        >
          <a
            href="https://www.sporran.org/terms"
            rel="noreferrer"
            role="menuitem"
            tabindex="-1"
            target="_blank"
          >
            Terms & Conditions
          </a>
        </li>
        <li
          class="listItem"
        >
          <a
            href="https://support.kilt.io/"
            rel="noreferrer"
            role="menuitem"
            tabindex="-1"
            target="_blank"
          >
            Tech Support
          </a>
        </li>
        <li
          class="listItem"
        >
          <a
            href="/access"
            role="menuitem"
            tabindex="-1"
          >
            Toggle Website Access
          </a>
        </li>
        <li
          class="listItem"
        >
          <a
            href="/settings"
            role="menuitem"
            tabindex="-1"
          >
            Custom Endpoint
          </a>
        </li>
        <li
          class="listItem"
        >
          <a
            role="menuitem"
            tabindex="-1"
          >
            Version 2023.4.12
          </a>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Settings should render 1`] = `
<div>
  <div
    class="wrapper"
  >
    <button
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="Settings"
      class="toggle"
      role="button"
      tabindex="0"
      title="Settings"
      type="button"
    />
  </div>
</div>
`;

exports[`Settings should render the endpoint item in the internal build 1`] = `
<div>
  <div
    class="wrapper"
  >
    <button
      aria-expanded="true"
      aria-haspopup="true"
      aria-label="Settings"
      class="toggle"
      role="button"
      tabindex="0"
      title="Settings"
      type="button"
    />
    <div
      class="dropdown"
      role="menu"
    >
      <h4
        class="heading"
      >
        Settings
      </h4>
      <ul
        class="list"
      >
        <li
          class="listItem"
        >
          <a
            href="https://www.sporran.org/terms"
            rel="noreferrer"
            role="menuitem"
            tabindex="-1"
            target="_blank"
          >
            Terms & Conditions
          </a>
        </li>
        <li
          class="listItem"
        >
          <a
            href="https://support.kilt.io/"
            rel="noreferrer"
            role="menuitem"
            tabindex="-1"
            target="_blank"
          >
            Tech Support
          </a>
        </li>
        <li
          class="listItem"
        >
          <a
            href="/access"
            role="menuitem"
            tabindex="-1"
          >
            Toggle Website Access
          </a>
        </li>
        <li
          class="listItem"
        >
          <a
            href="/settings"
            role="menuitem"
            tabindex="-1"
          >
            Custom Endpoint
          </a>
        </li>
        <li
          class="listItem"
        >
          <a
            role="menuitem"
            tabindex="-1"
          >
            Version 2023.4.12
          </a>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
